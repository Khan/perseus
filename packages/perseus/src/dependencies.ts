import * as React from "react";

import type {PerseusDependencies, PerseusDependenciesV2} from "./types";

let _dependencies: PerseusDependencies | null | undefined = null;

export const setDependencies = (dependencies: PerseusDependencies) => {
    _dependencies = dependencies;
};

export const getDependencies = (): PerseusDependencies => {
    if (_dependencies) {
        return _dependencies;
    }

    throw new Error(
        [
            "Perseus has not been provided required dependencies.",
            "setDependencies(dependencies) must be called first.",
            "Make sure Perseus is being imported from javascript/perseus/perseus.js.",
        ].join("\n"),
    );
};

export type DependencyProps = Partial<
    React.ContextType<typeof DependenciesContext>
>;

export const DependenciesContext = React.createContext<PerseusDependenciesV2>({
    analytics: {onAnalyticsEvent: async () => {}},
});

/**
 * useDependencies provides easy access to the current Perseus dependencies.
 */
export const useDependencies = (): PerseusDependenciesV2 => {
    const deps = React.useContext(DependenciesContext);

    return deps;
};
