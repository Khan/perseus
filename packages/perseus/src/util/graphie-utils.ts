import {Errors, PerseusError} from "@khanacademy/perseus-core";

import {getDependencies} from "../dependencies";
import {Log} from "../logging/log";
import Util from "../util";

import type {Coord} from "../interactive2/types";
import type {CSSProperties} from "aphrodite";

// For offline exercises in the mobile app, we download the graphie data
// (svgs and localized data files) and serve them from the local file
// system (with file://). We replace urls that start with `web+graphie`
// in the perseus json with this `file+graphie` prefix to indicate that
// they should have the `file://` protocol instead of `https://`.
const svgLocalLabelsRegex = /^file\+graphie:/;
const hashRegex = /\/([^/]+)$/;

/**
 * A Graphie axis range.
 */
export type GraphieRange = [number, number];

/**
 * A Graphie label.
 */
export type GraphieLabel = {
    // The text of the label.
    content: string;

    // The direction of the label. A number indicates rotation angle.
    alignment:
        | number
        | "center"
        | "above"
        | "above right"
        | "right"
        | "below right"
        | "below"
        | "below left"
        | "left"
        | "above left";

    // The label position in the x, y Graphie axis range.
    coordinates: GraphieRange;

    // Additional label styling.
    style: CSSProperties;

    // Whether content is TeX.
    typesetAsMath: boolean;
};

/**
 * The Graphie data (labels, etc.)
 */
export type GraphieData = {
    // The x, y axis range of the Graphie.
    range: [GraphieRange, GraphieRange];

    // The labels in the Graphie.
    labels: Array<GraphieLabel>;
};

function getLocale() {
    const {JIPT, kaLocale} = getDependencies();
    return JIPT.useJIPT ? "en-pt" : kaLocale;
}

function shouldUseLocalizedData() {
    return getLocale() !== "en";
}

// A regex to split at the last / of a URL, separating the base part from the
// hash. This is used to create the localized label data URLs.
const splitHashRegex = /\/(?=[^/]+$)/;

export function getLocalizedDataUrl(url: string) {
    // For local (cached) graphie images, they are already localized.
    if (svgLocalLabelsRegex.test(url)) {
        return Util.getDataUrl(url);
    }
    const [base, hash] = Util.getBaseUrl(url).split(splitHashRegex);
    return `${base}/${getLocale()}/${hash}-data.json`;
}

// Get the hash from the url, which is just the filename
function getUrlHash(url: string) {
    const match = url.match(hashRegex);
    if (match == null) {
        throw new PerseusError("not a valid URL", Errors.InvalidInput);
    }
    return match && match[1];
}

/**
 * Parse JSONP for a web+graphie://... hash to extract Graphie data, or throw
 * an error if invalid JSONP.
 */
export function parseDataFromJSONP(
    graphieHash: string,
    graphieJSONP: string,
    errorCallback: (error?: any) => void,
): GraphieData | null {
    // The JSONP is expected to be in the form of `svgDataHASH(...)` or `svgOtherDataHASH(...)`
    const match = graphieJSONP.match(
        new RegExp(`^(?:svgData|svgOtherData)${graphieHash}\\((.+)\\);$`),
    );

    // It is also possible that the JSONP is simply a JSON object,
    //  in which case we can parse it directly.
    const jsonToParse = match ? match[1] : graphieJSONP;

    // Try to parse the JSONP, and if it fails, call the error callback
    try {
        return JSON.parse(jsonToParse);
    } catch (error) {
        errorCallback(error);
        return null;
    }
}

type CacheEntry = {
    labels: ReadonlyArray<any>;
    range: [Coord, Coord];
};

// The global cache of label data. Its format is:
// {
//   hash (e.g. "c21435944d2cf0c8f39d9059cb35836aa701d04a"): {
//     loaded: a boolean of whether the data has been loaded or not
//     dataCallbacks: a list of callbacks to call with the data when the data
//                    is loaded
//     data: the other data for this hash
//   },
//   ...
// }
const labelDataCache: Record<
    string,
    | {
          loaded: false;
          localized: boolean;
          dataCallbacks: Array<any>;
      }
    | {
          loaded: true;
          data: CacheEntry;
          localized: boolean;
          dataCallbacks: Array<any>;
      }
> = {};

export function loadGraphie(
    url: string,
    onDataLoaded: (data: CacheEntry, localized: boolean) => void,
) {
    const hash = getUrlHash(url);

    // We can't make multiple jsonp calls to the same file because their
    // callbacks will collide with each other. Instead, we cache the data
    // and only make the jsonp calls once.
    const entry = labelDataCache[hash];
    if (entry != null) {
        if (entry.loaded) {
            const {data, localized} = entry;
            onDataLoaded(data, localized);
        } else {
            entry.dataCallbacks.push(onDataLoaded);
        }
    } else {
        const cacheData = {
            loaded: false as const,
            dataCallbacks: [onDataLoaded],
            localized: shouldUseLocalizedData(),
        };

        labelDataCache[hash] = cacheData;

        const retrieveData = async (
            url: string,
            errorCallback: (x?: any, status?: any, error?: any) => void,
        ) => {
            const response = await fetch(url);

            if (!response?.ok) {
                errorCallback();
                return;
            }

            const jsonp = await response.text();

            const data = parseDataFromJSONP(hash, jsonp, (error) => {
                Log.error(
                    "Failed to parse JSONP for svg-image",
                    Errors.Service,
                    {
                        cause: error,
                        loggedMetadata: {
                            dataUrl: Util.getDataUrl(url),
                            jsonp,
                        },
                    },
                );
            });

            // If the data is null, then we failed to parse the JSONP
            if (!data) {
                return;
            }

            const newCacheEntry = (labelDataCache[hash] = {
                ...labelDataCache[hash],
                loaded: true as const,
                data,
            });

            newCacheEntry.dataCallbacks.forEach((callback) => {
                callback(newCacheEntry.data, cacheData.localized);
            });
        };

        const dataLoadErrorHandler = (x, status, error) => {
            Log.error("Data load failed for svg-image", Errors.Service, {
                cause: error,
                loggedMetadata: {
                    dataUrl: Util.getDataUrl(url),
                    status,
                },
            });
        };

        if (shouldUseLocalizedData()) {
            cacheData.localized = false;
            // If there is isn't any localized data, use the error callback to try to
            // load the non-localized data. If that fails too, then we will throw an error.
            retrieveData(getLocalizedDataUrl(url), () =>
                retrieveData(Util.getDataUrl(url), dataLoadErrorHandler),
            );
        } else {
            retrieveData(Util.getDataUrl(url), dataLoadErrorHandler);
        }
    }
}
