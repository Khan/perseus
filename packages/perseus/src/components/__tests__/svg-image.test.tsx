import {act, render} from "@testing-library/react";
import * as React from "react";

import {testDependencies} from "../../../../../testing/test-dependencies";
import * as Dependencies from "../../dependencies";
import {typicalCase} from "../../util/graphie-utils.testdata";
import SvgImage from "../svg-image";

describe("SvgImage", () => {
    const images: Array<Record<any, any>> = [];
    let originalImage;

    beforeEach(() => {
        jest.clearAllMocks();
        originalImage = window.Image;
        // Mock HTML Image so we can trigger onLoad callbacks and see full
        // image rendering.
        // @ts-expect-error - TS2322 - Type 'Mock<Record<string, any>, [], any>' is not assignable to type 'new (width?: number | undefined, height?: number | undefined) => HTMLImageElement'.
        // TODO(LEMS-3083): Remove eslint suppression
        // eslint-disable-next-line functional/immutable-data
        window.Image = jest.fn(() => {
            const img: Record<string, any> = {};
            // TODO(LEMS-3083): Remove eslint suppression
            // eslint-disable-next-line functional/immutable-data
            images.push(img);
            return img;
        });

        // TODO(LEMS-3083): Remove eslint suppression
        // eslint-disable-next-line functional/immutable-data
        global.fetch = jest.fn((url) => {
            return Promise.resolve({
                text: () => Promise.resolve(typicalCase.jsonpString),
                ok: true,
            });
        }) as jest.Mock;
    });

    afterEach(() => {
        // TODO(LEMS-3083): Remove eslint suppression
        // eslint-disable-next-line functional/immutable-data
        window.Image = originalImage;
    });

    // Tells the image loader 1, or all, of our images loaded
    const markImagesAsLoaded = (imageIndex?: number) => {
        if (imageIndex != null) {
            const img = images[imageIndex];
            if (img?.onload) {
                act(() => img.onload());
            }
        } else {
            images.forEach((i) => {
                if (i?.onload) {
                    act(() => i.onload());
                }
            });
        }
    };
    it("should render a spinner initially", () => {
        // Arrange
        jest.spyOn(Dependencies, "getDependencies").mockReturnValue(
            testDependencies,
        );

        // Act
        const {container} = render(
            <SvgImage src="http://localhost/sample.png" alt="png image" />,
        );

        // Assert
        expect(
            // eslint-disable-next-line testing-library/no-container, testing-library/no-node-access
            container.querySelector("div[class*=spinnerContainer]"),
        ).toBeVisible();
        expect(container).toMatchSnapshot();
    });

    it("should load and render a png", () => {
        // Arrange
        jest.spyOn(Dependencies, "getDependencies").mockReturnValue(
            testDependencies,
        );

        // Act
        const {container} = render(
            <SvgImage src="http://localhost/sample.png" alt="png image" />,
        );

        markImagesAsLoaded(); // Tell the ImageLoader that our images are loaded

        // Assert
        expect(container).toMatchSnapshot();
    });

    it("should load and render a normal graphie svg", async () => {
        // Arrange
        jest.spyOn(Dependencies, "getDependencies").mockReturnValue(
            testDependencies,
        );

        // Act
        const {container} = render(
            <SvgImage src={typicalCase.url} alt="svg image" />,
        );

        markImagesAsLoaded(); // Tell the ImageLoader that our images are loaded

        // Assert
        expect(container).toMatchSnapshot();
    });

    it("should load and render a localized graphie svg", async () => {
        // Arrange
        jest.spyOn(Dependencies, "getDependencies").mockReturnValue({
            ...testDependencies,
            kaLocale: "es",
        });

        // Act
        const {container} = render(
            <SvgImage src={typicalCase.url} alt="svg image" />,
        );

        markImagesAsLoaded(); // Tell the ImageLoader that our images are loaded

        // Assert
        expect(container).toMatchSnapshot();
    });
});
