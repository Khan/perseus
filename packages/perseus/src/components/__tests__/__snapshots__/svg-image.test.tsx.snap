// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`SvgImage should load and render a localized graphie svg 1`] = `
<div>
  <div
    class="unresponsive-svg-image"
  >
    <img
      alt="svg image"
      src="mockStaticUrl(https://ka-perseus-graphie.s3.amazonaws.com/ccefe63aa1bd05f1d11123f72790a49378d2e42b.svg)"
    />
  </div>
</div>
`;

exports[`SvgImage should load and render a normal graphie svg 1`] = `
<div>
  <div
    class="unresponsive-svg-image"
  >
    <img
      alt="svg image"
      src="mockStaticUrl(https://ka-perseus-graphie.s3.amazonaws.com/ccefe63aa1bd05f1d11123f72790a49378d2e42b.svg)"
    />
  </div>
</div>
`;

exports[`SvgImage should load and render a png 1`] = `
<div>
  <img
    alt="png image"
    src="mockStaticUrl(http://localhost/sample.png)"
  />
</div>
`;

exports[`SvgImage should render a spinner initially 1`] = `
<div>
  <span
    style="top: 0px; left: 0px; width: 100%; height: 100%; position: absolute; min-width: 20px; display: flex; justify-content: center; align-content: center;"
  >
    <div
      class="default_xu2jcg-o_O-spinnerContainer_agrn11"
    >
      <svg
        height="48"
        viewBox="0 0 48 48"
        width="48"
        xmlns="http://www.w3.org/2000/svg"
      >
        <path
          class="loadingSpinner_1dvgeb3-o_O-inlineStyles_8njutv"
          d="M44.19 23.455a1.91 1.91 0 1 1 3.801 0h.003c.004.18.006.363.006.545 0 13.255-10.745 24-24 24S0 37.255 0 24 10.745 0 24 0c.182 0 .364.002.545.006V.01a1.91 1.91 0 1 1 0 3.801v.015A20.564 20.564 0 0 0 24 3.818C12.854 3.818 3.818 12.854 3.818 24c0 11.146 9.036 20.182 20.182 20.182 11.146 0 20.182-9.036 20.182-20.182 0-.182-.003-.364-.007-.545h.015z"
          fill-rule="nonzero"
        />
      </svg>
    </div>
  </span>
</div>
`;
